File: src/CommerceRuntime/Contoso.Commerce.Runtime.MarginCalculation/GetMarginCalculationRequest.cs
// ----------------------------------------------------------------------------
// Copyright (c) Contoso. All rights reserved.
// ----------------------------------------------------------------------------

namespace Contoso.Commerce.Runtime.MarginCalculation
{
    using System.Runtime.Serialization;
    using Microsoft.Dynamics.Commerce.Runtime.Messages;

    /// <summary>
    /// Request sent from POS to the CRT extension to calculate the margin
    /// for a selected sales order line.
    /// </summary>
    [DataContract]
    public sealed class GetMarginCalculationRequest : Request
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="GetMarginCalculationRequest"/> class.
        /// </summary>
        /// <param name="itemId">The item identifier.</param>
        /// <param name="quantity">Quantity on the sales line.</param>
        /// <param name="netAmount">Net amount (revenue) on the sales line.</param>
        /// <param name="dataAreaId">The legal entity / company identifier (e.g. "USMF").</param>
        public GetMarginCalculationRequest(string itemId, decimal quantity, decimal netAmount, string dataAreaId)
        {
            this.ItemId = itemId;
            this.Quantity = quantity;
            this.NetAmount = netAmount;
            this.DataAreaId = dataAreaId;
        }

        /// <summary>Gets the item identifier.</summary>
        [DataMember]
        public string ItemId { get; private set; }

        /// <summary>Gets the quantity on the sales line.</summary>
        [DataMember]
        public decimal Quantity { get; private set; }

        /// <summary>
        /// Gets the net amount of the sales line – used as the revenue figure
        /// in the margin formula:  Margin% = (Revenue − Cost) / Revenue × 100.
        /// </summary>
        [DataMember]
        public decimal NetAmount { get; private set; }

        /// <summary>Gets the legal-entity identifier used to look up the purchase price.</summary>
        [DataMember]
        public string DataAreaId { get; private set; }
    }
}

File: src/Pos/MarginCalculationExtension/Messages/GetMarginCalculationRequest.ts
// ----------------------------------------------------------------------------
// Copyright (c) Contoso. All rights reserved.
// ----------------------------------------------------------------------------

/**
 * Data class for passing margin calculation parameters to the CRT service.
 * Maps to C# GetMarginCalculationRequest in the CRT extension.
 *
 * Note: In Retail SDK 7.2.x, Commerce.Proxy.Common does not exist.
 * This is a plain TypeScript class used to hold the request parameters.
 * The actual server call is made from MarginCalculationOperation.ts via
 * the Commerce proxy manager once the Retail Server extension is deployed.
 */
export class GetMarginCalculationRequest {
    /** Item identifier selected on the sales line. */
    public itemId: string;

    /** Quantity on the selected sales line. */
    public quantity: number;

    /**
     * Net amount of the selected sales line.
     * Used as the revenue figure in the margin formula:
     * Margin % = (NetAmount - PurchasePrice x Qty) / NetAmount x 100
     */
    public netAmount: number;

    /**
     * Legal entity / company identifier (e.g. "USMF").
     * Passed to the D365 F&O real-time service to scope the purchase-price lookup.
     */
    public dataAreaId: string;

    constructor(itemId: string, quantity: number, netAmount: number, dataAreaId: string) {
        this.itemId = itemId;
        this.quantity = quantity;
        this.netAmount = netAmount;
        this.dataAreaId = dataAreaId;
    }
}
